cmd/provider/mirror.go

```diff
@@ -0,0 +1,239 @@
+		Name:        "storePath",
+		Usage:       "The path at which to persist the mirror state.",
+		DefaultText: "Ephemeral in-memory storage",
+	}
+	Mirror.flags.initAdRecurLimit = &cli.UintFlag{
+		Name:        "initAdRecurLimit",
+		Usage:       "The maximum recursion depth limit of ads to mirror if no previous ads are mirrored.",
+		DefaultText: "No limit",
+	}
+	Mirror.flags.entriesRecurLimit = &cli.UintFlag{
+		Name:        "entriesRecurLimit",
+		Usage:       "The maximum recursion depth limit of ad entries to mirror.",
+		DefaultText: "No limit",
+	}
+	Mirror.flags.remapWithEntryChunkSize = &cli.UintFlag{
+		Name:        "remapWithEntryChunkSize",
+		Usage:       "Remaps the advertisement entries to EntryChunk chain with the specified chunk size",
+		DefaultText: "No remapping of entries",
+	}
+	Mirror.flags.remapWithHamtHashFunc = &cli.StringFlag{
+		Name: "remapWithHamtHashFunc",
+		Usage: "Remaps the advertisement entries to HAMT using the given hash function. " +
+			"Only `identity`, `sha2-256` and `murmur3-x64-64` are accepted.",
+		DefaultText: "No remapping of entries",
+	}
+	Mirror.flags.remapWithHamtBitWidth = &cli.UintFlag{
+		Name:        "remapWithHamtBitWidth",
+		Usage:       "Remaps the advertisement entries to HAMT using the given bit-width.",
+		DefaultText: "No remapping of entries",
+	}
+	Mirror.flags.remapWithHamtBucketSize = &cli.UintFlag{
+		Name:        "remapWithHamtBucketSize",
+		Usage:       "Remaps the advertisement entries to HAMT using the given bucket size.",
+		DefaultText: "No remapping of entries",
+	}
+	Mirror.flags.topic = &cli.StringFlag{
+		Name:        "topic",
+		Usage:       "The topic on which the source and mirrored advertisements are announced.",
+		DefaultText: "`/indexer/ingest/mainnet`",
+	}
+	Mirror.flags.skipRemapOnEntriesTypeMatch = &cli.BoolFlag{
+		Name:        "skipRemapOnEntriesTypeMatch",
+		Usage:       "Whether to skip remapping the entries if the source entries kind matches the required mirrored remap kind.",
+		DefaultText: "No remapping of entries",
+	}
+	Mirror.flags.alwaysReSignAds = &cli.BoolFlag{
+		Name:        "alwaysReSignAds",
+		Usage:       "Whether to always re-sign advertisements with the mirror's identity.",
+		DefaultText: "Ads are only re-singed if changed by the mirror.",
+	}
+	Mirror.cmd = &cli.Command{
+		Name:  "mirror",
+		Usage: "Mirrors the advertisement chain from an existing index provider.",
+		Flags: []cli.Flag{
+			Mirror.flags.source,
+			Mirror.flags.syncInterval,
+			Mirror.flags.identityPath,
+			Mirror.flags.listenAddrs,
+			Mirror.flags.storePath,
+			Mirror.flags.initAdRecurLimit,
+			Mirror.flags.entriesRecurLimit,
+			Mirror.flags.remapWithEntryChunkSize,
+			Mirror.flags.remapWithHamtHashFunc,
+			Mirror.flags.remapWithHamtBitWidth,
+			Mirror.flags.remapWithHamtBucketSize,
+			Mirror.flags.topic,
+			Mirror.flags.skipRemapOnEntriesTypeMatch,
+			Mirror.flags.alwaysReSignAds,
+		},
+		Before: func(cctx *cli.Context) error {
```

Comment:
does this func need to be in `init`?

Reply:
Not really. Happy to move it out if you think it makes the code harder to read.

Reply:
I'd have these methods defined at the top level, rather than inline.